plan/physical_plan_builder.go

```diff
@@ -571,6 +572,228 @@ func (p *Join) convert2PhysicalPlanRight(prop *requiredProperty, innerJoin bool)
+		for _, cond := range p.RightConditions {
+			newCond := convertCol2CorCol(cond, corCols, outerSchema)
+			conds = append(conds, newCond)
+		}
+	} else {
+		conds = make([]expression.Expression, 0, len(p.EqualConditions)+len(p.LeftConditions)+len(p.OtherConditions))
+		for _, cond := range p.LeftConditions {
+			newCond := convertCol2CorCol(cond, corCols, outerSchema)
+			conds = append(conds, newCond)
+		}
+	}
+	for _, cond := range p.EqualConditions {
+		newCond := convertCol2CorCol(cond, corCols, outerSchema)
+		conds = append(conds, newCond)
+	}
+	for _, cond := range p.OtherConditions {
+		newCond := convertCol2CorCol(cond, corCols, outerSchema)
+		conds = append(conds, newCond)
+	}
+	selection.Conditions = conds
+	_, selection.canControlScan = selection.makeScanController(true)
+	return selection, corCols
+}
+
+func (p *Join) convert2IndexNestedLoopJoinLeft(prop *requiredProperty, innerJoin bool) (*physicalPlanInfo, error) {
+	lChild := p.children[0].(LogicalPlan)
+	if _, ok := p.children[1].(*DataSource); !ok {
+		return nil, nil
+	}
+	allLeft := true
+	for _, col := range prop.props {
+		if !lChild.Schema().Contains(col.col) {
+			allLeft = false
+		}
+	}
+	lProp := prop
+	if !allLeft {
+		lProp = &requiredProperty{}
+	} else {
+		lProp = replaceColsInPropBySchema(prop, lChild.Schema())
+	}
+	var lInfo *physicalPlanInfo
+	var err error
+	if innerJoin {
+		lInfo, err = lChild.convert2PhysicalPlan(removeLimit(lProp))
+	} else {
+		lInfo, err = lChild.convert2PhysicalPlan(convertLimitOffsetToCount(lProp))
+	}
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if lInfo.p == nil {
+		return nil, nil
+	}
+	selection, corCols := p.buildSelectionWithConds(true)
+	if !selection.canControlScan {
+		return nil, nil
+	}
+	rInfo, _ := selection.makeScanController(false)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	join := &PhysicalHashJoin{
+		LeftConditions: p.LeftConditions,
+		// TODO: decide concurrency by data size.
+		Concurrency:   JoinConcurrency,
+		DefaultValues: p.DefaultValues,
+		SmallTable:    1,
+	}
+	join.tp = "NestedLoopJoinLeft"
+	join.allocator = p.allocator
+	join.initIDAndContext(p.ctx)
+	join.SetChildren(lInfo.p, rInfo.p)
+	if innerJoin {
+		join.JoinType = InnerJoin
+	} else {
+		join.JoinType = LeftOuterJoin
+	}
+	join.SetSchema(p.Schema())
+	resultInfo := join.matchProperty(prop, lInfo, rInfo)
+	ap := &PhysicalApply{
+		PhysicalJoin: resultInfo.p,
+		OuterSchema:  corCols,
+	}
+	ap.tp = App
+	ap.allocator = p.allocator
+	ap.initIDAndContext(p.ctx)
+	ap.SetChildren(resultInfo.p.Children()...)
+	ap.SetSchema(resultInfo.p.Schema())
+	resultInfo.p = ap
+	if !allLeft {
+		resultInfo = enforceProperty(prop, resultInfo)
+	} else {
+		resultInfo = enforceProperty(limitProperty(prop.limit), resultInfo)
+	}
+	return resultInfo, nil
+}
+
+func (p *Join) convert2IndexNestedLoopJoinRight(prop *requiredProperty, innerJoin bool) (*physicalPlanInfo, error) {
```

Comment:
These two functions seem almost the same. Can we try make it one gracefully?
Very often ppl fix bug one place but ignored the copied version.